Subproblem callbacks

junglegobs opened the pull request:
Checklist so far:

* [x] Created SPCB structs, with documentation + checked that they work
* [x] Added vector of SPCB to `solve`
* [x] Added vector of SPCB to `initialize`
* [x] Added SPCB to `_initialize_subproblems`
* [x] Added SPCB to `Initialize` message 
* [x] Added callbacks to subproblem records
* [x] Created `_execute_subproblem_callback` function(s)
* [x] Added `niter` to `Solve` message
* [ ] Create tests and check if works (haven't done that yet)
* [ ] Added callback to progressively decrease an optimizer parameter (e.g. MIP gap)

junglegobs commented:
@jmaack24 For some reason the algorithm "hangs" when processing `Solve`:
```julia
ERROR: LoadError: InterruptException:
Stacktrace:
 [1] try_yieldto(::typeof(Base.ensure_rescheduled)) at ./task.jl:656
 [2] wait at ./task.jl:713 [inlined]
 [3] wait(::Base.GenericCondition{Base.Threads.SpinLock}) at ./condition.jl:106
 [4] _wait(::Task) at ./task.jl:238
 [5] sync_end(::Channel{Any}) at ./task.jl:294
 [6] macro expansion at ./task.jl:333 [inlined]
 [7] _send_solve_commands(::ProgressiveHedging.PHData, ::ProgressiveHedging.WorkerInf, ::Int64) at /home/u0128861/Desktop/Julia_Pkg_Dev/dev/ProgressiveHedging.jl/src/algorithm.jl:84
```
This is weird, because removing my command, `_execute_subproblem_callbacks(output, record, msg.niter)` does not fix this, only having an empty `SubproblemCallback` array does. I'm also unable to get print statements to work at the worker level, so I can't really debug.

jmaack24 commented:
I noticed there are two `_execute_subproblem_callbacks` calls.  Did you take both out?

It also seems like you don't need the one during initialization.  Anything the user wants to do with callbacks there can easily be done during subproblem creation by the subproblem creation function or during the first execution of the solve message.

Back to why it's hanging. I can think of two things that would cause that. The master and the worker both waiting for something (i.e. the master is waiting for a reply and the worker is waiting for a command).  The other is the master waiting to put something on the worker queue because it is full and the worker waiting to put something on the master queue because it is also full.  I tried to size the queues (channels actually) so that the second can't happen.  

If the print statements aren't working, then that likely means the worker is hanging before it gets to the print statements.  You can also put print statements on the master side to see what it's doing.

junglegobs commented:
Fixed the hanging issue (I think it was due to `_execute_subproblem_callbacks` receiving `output` as an argument) as well as:
- Removed the "intiialization" function.
- Changed function call from using the `AbstractSubproblem` to the `SubproblemRecord`. I think there's not efficiency loss from doing this at the worker level, and it gives the user access to more information to use in their function. It might be worth exporting `SubproblemRecord` in that case though.

From my very lazy test, I noticed that the same callback is called 4 times for the same PH iteration and scenario id - perhaps this is normal since you have a 3 stage problem, but at first glance this strikes me as wrong?

At this point I should probably write some tests, but from checking what you did for the callback functions this might be quite involved. Would you mind doing that? Otherwise you can tell me what to do and I'll do it.

jmaack24 commented on src/algorithm.jl:
```diff
@@ -79,10 +79,15 @@ end
     @sync for (scen, sinfo) in pairs(phd.scenario_map)
         (w_dict, xhat_dict) = create_ph_dicts(sinfo)
-        @async _send_message(winf, sinfo.pid, Solve(scen, w_dict, xhat_dict))
+        @async _send_message(winf, sinfo.pid, Solve(
```
Reformat here to be consistent with PH coding:
```julia
@async _send_message(winf,
                     sinfo.pid,
                     Solve(scen,
                           w_dict,
                           xhat_dict,
                           niter)
                     )
```

jmaack24 commented on src/structs.jl:
```diff
@@ -247,6 +247,17 @@ end
+*`indexer::Indexer`
+*`iterate_history::PHIterateHistory`
+*`residual_history::PHResidualHistory`
+*`time_info::TimerOutputs.TimerOutput`
```
This does need some documentation.  I think a better way to go is to refer to the available functions for accessing things.  Something like
```
...
For interacting with this object see: [`consensus_variables`](@ref)
...
```

The idea is that users shouldn't (for the most part) need to reach into the internals of the `PHData` object but interact with it through functions.  That way we do not expose internals that the user doesn't need and could break.

jmaack24 replied:
Also the two built in Julia functions `methods` and `methodswith` are extremely helpful in these situations.

jmaack24 commented on src/subproblem_callback.jl:
```diff
@@ -0,0 +1,57 @@
+"""
+    spcb(f::Function)
+
+Shorthand for `SubproblemCallback(f)`.
```
Is it worth adding the shorthand functions for all the forms? Also a question for normal callbacks.

junglegobs replied:
In my opinion no, though I suppose `SubproblemCallback` is a bit laborious to type. I think it doesn't hurt.

jmaack24 commented on src/subproblem_callback.jl:
```diff
@@ -0,0 +1,57 @@
+* `ext::Dict{Symbol,Any}` : Dictionary to store data between callback calls or needed parameters.
+
+The callback function `h` must have the signature
+    `h(ext::Dict{Symbol,Any}, sp::SubproblemRecord, niter::Int, scenario_id::ScenarioID)::Bool`
```
A couple of things here.  First, it doesn't seem like the subproblem callback needs to return anything. The return value isn't used at all so you can leave it entirely unspecified and remove one thing for the user to worry about.  For the normal callbacks, the return value allows the user to terminate PH for any reason they see fit.  Here we don't have that need.

Second, the subproblem record is meant to contain things that PH needs to run but that the user doesn't need to know anything about.  That's why they are stored in a separate data struct and not in the subproblem itself (it's a data encapsulation issue).  If the user changes these data fields (on purpose or accidentally), PH could break in very unusual and non-obvious ways.  So the `SubproblemRecord` should not be exposed to the user and only the subproblem should be handed to the callback.

junglegobs replied:
Ah ok, I was wondering about that. Makes sense, will fix accordingly.

jmaack24 commented on src/subproblem_callback.jl:
```diff
@@ -0,0 +1,57 @@
+"""
+function SubproblemCallback(f::Function, ext::Dict{Symbol,Any}, name::String)
+    return SubproblemCallback(string(f), f, ext, name)
+end
```
All files should end with a black line.  Please add a new line here.  I'm actually surprised your editor doesn't do it automatically.

jmaack24 commented on src/worker.jl:
```diff
@@ -260,3 +267,15 @@ function _split_variables(scen_tree::ScenarioTree,
+                                       )::Nothing 
+    for (scen, sub) in record.subproblems
+        for spcb in sub.subproblem_callbacks
+            spcb.h(spcb.ext, sub, niter, scen)
```
Again, this should just be the subproblem and not the full subproblem record.

jmaack24 commented on test/test_setup.jl:
```diff
@@ -31,7 +31,7 @@
 
         PH._build_subproblems(rc, wr, PH.scenarios(st), st,
-                              create_model, (), NamedTuple())
+                              create_model, (), NamedTuple(), SubproblemCallback[])
```
More unusual formatting.

junglegobs replied:
I'm not sure what you mean by unusual formatting here - it's the same as before but with the additional `SubproblemCallback[]`.

jmaack24 replied:
Nevermind. Looks like github displayed it in an unusual way.  Sorry about that.

jmaack24 commented on test/test_subproblem_callback.jl:
```diff
@@ -0,0 +1,26 @@
+                subproblem_callbacks=[SubproblemCallback(my_subproblem_callback)]
+            )
+    
+end
```
New line at end of file again.

jmaack24 commented on src/worker.jl:
```diff
@@ -129,6 +131,8 @@ function process_message(msg::Solve,
 
     update_ph_terms(sub.problem, msg.w_vals, msg.xhat_vals)
 
+    _execute_subproblem_callbacks(record, msg.niter)
```
This gets called for every solve message that is received.  There is a solve message for each scenario.  `execute_subproblem_callbacks` seems to execute all callbacks on all subproblems.  This is why it gets called 12 times total.

jmaack24 commented on src/worker.jl:
```diff
@@ -260,3 +267,15 @@ function _split_variables(scen_tree::ScenarioTree,
+function _execute_subproblem_callbacks(record::WorkerRecord,
+                                       niter::Int
+                                       )::Nothing 
+    for (scen, sub) in record.subproblems
```
Shouldn't run over all subproblems. See above comment in `process_message(msg::Solve...`

jmaack24 commented:
> From my very lazy test, I noticed that the same callback is called 4 times for the same PH iteration and scenario id - perhaps this is normal since you have a 3 stage problem, but at first glance this strikes me as wrong?

If it gets called 12 times total (3 times for each scenario), then something is wrong.  If it gets called 4 times (once per scenario), then it is fine.

> At this point I should probably write some tests, but from checking what you did for the callback functions this might be quite involved. Would you mind doing that? Otherwise you can tell me what to do and I'll do it.

The callback test were somewhat difficult because of where they're located.  You can rig up something that just calls `_execute_subproblem_callbacks` once and check that the callback you put in the subproblem record gets called.  Something simple like adding some string or number to the `ext`.  Perhaps also do something with the subproblem in the callback to make sure that is passed in correctly.

junglegobs commented on src/setup.jl:
```diff
@@ -21,7 +21,8 @@ function _initialize_subproblems(sp_map::Dict{Int,Set{ScenarioID}},
                                  r::AbstractPenaltyParameter,
-                                 warm_start::Bool;
+                                 warm_start::Bool,
+                                 subproblem_callbacks::Vector{SubproblemCallback};
```
Yes, I use tabs. I have to admit I don't know how you format PH.jl - I use VSCode and it doesn't seem to work well with your formatting, perhaps there's a plugin I should be using?

jmaack24 replied:
I mostly just go by what "looks good" which is totally subjective and unhelpful.  But some rough guidelines:

1. Use whitespace (blank lines, separating long lines, etc.) liberally.  This makes the code easier on the eyes.
2. A single line shouldn't be more than roughly 80 characters (with the notable exception of string literals).
3. If a function signature or call doesn't fit on a single line break it up with the first argument following the function, and every other argument on its own line. I've started doing this in the last year. It is not consistently done throughout PH because of that.  Feel free to update areas with this guideline.
4. Within a file, group functions in "logical" groups. Try to alphabetize within a group.  The most obvious groupings are internal functions (i.e. not meant to be exposed to the user--normally preceeded by an "_" in the name) and external functions.

I hope that helps a little bit. I should probably post some sort of guidelines but before your contributions, I was the only one working in this code and it wasn't necessary.

There does seem to be a julia extension for VSCode: https://github.com/julia-vscode/julia-vscode

I just googled "julia vscode" and. that was what first came up.  There may be other options you like better.

junglegobs replied:
I mostly follow 1, 2 and 3, but my formatting is usually slightly different, e.g.
```julia
    param = Dict(
        "optimizer" => get_optimizer(sim),
        "relativePathTimeSeriesCSV" => ts_file,
        "optimizationHorizon" => Dict(
            "start" => [1,1,1],
            "end" => [
                1,
                length(sim["years to consider"])*length(sim["periods to consider"]),
                sim["time steps per period"]
            ]
        ),
    )
```
or
```julia

function build_GEP_scenario(
        scenario_id::PH.ScenarioID, 
        param::Dict,
        configFile,
        sim::Simulation,
        geppr_ts_file::String
    )
```
Atom (my previous editor) and VSCode allow for this quite easily, whereas your formatting is a little less convenient since I have to tab and then space to the opening bracket `(`, and similarly for the closing bracket.

I'll try to follow your guideline also for 4 next time I work on this tomorrow. On a sidenote, I think that `test_subproblem_callbacks` is probably unnecessary, since it's just two tests.

jmaack24 replied:
I use emacs which does both automatically for me.  Your first example is actually exactly why I use this formatting.  I find it significantly easier to see what's going on with this formatting
```julia
param = Dict("optimizer" => get_optimizer(sim),
             "relativePathTimeSeriesCSV" => ts_file,
             "optimizationHorizon" => Dict("start" => [1,1,1],
                                           "end" => [1,
                                                     length(sim["years to consider"])*length(sim["periods to consider"]),
                                                     sim["time steps per period"]
                                                     ]
                                           ),
             )
```
even with the line length.

Admittedly, its just because I've been looking at code like that for a long time and I'm just used to it.

junglegobs replied:
Found what I was looking for :-) https://marketplace.visualstudio.com/items?itemName=LAK132.indent-to-bracket

jmaack24 reviewed (approved):
Looks good.
